crates/llm-chain-sagemaker-endpoint/src/executor.rs

```diff
@@ -74,43 +77,120 @@ impl llm_chain::traits::Executor for Executor {
 
     fn tokens_used(
         &self,
-        _options: &Options,
-        _prompt: &Prompt,
+        options: &Options,
+        prompt: &Prompt,
     ) -> Result<TokenCount, PromptTokensError> {
-        // Not all models expose this information.
-        unimplemented!();
+        let tokenizer = self.get_tokenizer(options)?;
+        let input = prompt.to_text();
+
+        let tokens_used = tokenizer
+            .tokenize_str(&input)
+            .map_err(|_e| PromptTokensError::UnableToCompute)?
+            .len() as i32;
+        let max_tokens = self.max_tokens_allowed(options);
+        Ok(TokenCount::new(max_tokens, tokens_used)) 
     }
 
-    fn max_tokens_allowed(&self, _: &Options) -> i32 {
-        // Not all models expose this information.
-        unimplemented!();
+    fn max_tokens_allowed(&self, options: &Options) -> i32 {
+        let opts = self.cascade(Some(options));
+        let model = self.get_model_from_invocation_options(&opts);
+        model.context_window_size().unwrap_or_else(|| unimplemented!("This model does not expose max token allowed information."))
     }
 
     fn answer_prefix(&self, _prompt: &Prompt) -> Option<String> {
-        // Not all models expose this information.
-        unimplemented!();
+        None
     }
 
-    fn get_tokenizer(&self, _: &Options) -> Result<Self::StepTokenizer<'_>, TokenizerError> {
-        // Not all models expose this information.
-        unimplemented!();
+    fn get_tokenizer(&self, options: &Options) -> Result<Self::StepTokenizer<'_>, TokenizerError> {
+        Ok(SageMakerEndpointTokenizer::new(self.cascade(Some(options))))
     }
 }
 
-pub struct SageMakerEndpointTokenizer {}
+pub struct SageMakerEndpointTokenizer {
+    tokenizer: Option<HuggingFaceTokenizer>
+}
 
 impl SageMakerEndpointTokenizer {
-    pub fn new(_executor: &Executor) -> Self {
-        SageMakerEndpointTokenizer {}
+    pub fn new(options: OptionsCascade) -> Self {
+        let optional_tokenizer = match options.get(llm_chain::options::OptDiscriminants::Model) {
+            Some(Opt::Model(model)) => {
+               let model_struct = Model::from_str(&model.to_name()).unwrap(); 
+                Some(HuggingFaceTokenizer::from_pretrained(&model_struct.to_huggingface_name(), None).unwrap()) // TODO: no options
+            }
+            _ => None,
+        };
+        
+        SageMakerEndpointTokenizer {
+            tokenizer: optional_tokenizer
+        }
     }
 }
 
 impl Tokenizer for SageMakerEndpointTokenizer {
-    fn tokenize_str(&self, _doc: &str) -> Result<TokenCollection, TokenizerError> {
-        unimplemented!();
+    fn tokenize_str(&self, doc: &str) -> Result<TokenCollection, TokenizerError> {
+        match &self.tokenizer {
+            Some(tokenizer) => {
+                let encoding = tokenizer.encode(doc, false).map_err(|_| TokenizerError::TokenizationError)?;
+                let ids: Vec<i32> = encoding.get_ids().iter().map(|x| *x as i32).collect();
```

Comment:
When you need to provide a type hint for collect, it only cares about the container type so this is fine:
```suggestion
                let ids: Vec<_> = encoding.get_ids().iter().map(|x| *x as i32).collect();
```

Reply:
Thank you! Fixed.